mypy/semanal.py

```diff
@@ -1288,6 +1291,126 @@ def store_declared_types(self, lvalue: Node, typ: Type) -> None:
             # This has been flagged elsewhere as an error, so just ignore here.
             pass
 
+    def process_newtype_declaration(self, s: AssignmentStmt) -> None:
+        """Check if s declares a NewType; if yes, store it in symbol table."""
+        # Extract and check all information from newtype declaration
+        call = self.get_newtype_declaration(s)
+        if call is None:
+            return
+        call.analyzed = NewTypeExpr(None).set_line(call.line)
+
+        name = self.get_newtype_name(s)
+        if name is None:
+            return
+
+        old_type = self.check_newtype_args(call, name, s)
+        if old_type is None:
+            return
+
+        # Create the corresponding class definition if the aliased type is subtypeable
+        if isinstance(old_type, TupleType):
+            newtype_class_info = self.build_newtype_typeinfo(name, old_type, old_type.fallback)
+            newtype_class_info.tuple_type = old_type
+        elif isinstance(old_type, Instance):
+            newtype_class_info = self.build_newtype_typeinfo(name, old_type, old_type)
+        else:
+            message = "Argument 2 to NewType(...) must be subclassable (got {})"
+            self.fail(message.format(old_type), s)
+            return
+
+        # If so, add it to the symbol table.
+        node = self.lookup(name, s)
+        # TODO: why does NewType work in local scopes despite always being of kind GDEF?
+        node.kind = GDEF
+        node.node = newtype_class_info
+        call.analyzed = NewTypeExpr(newtype_class_info).set_line(call.line)
+
+    def get_newtype_declaration(self, s: AssignmentStmt) -> Optional[CallExpr]:
+        """Returns the Newtype() call statement if `s` is a newtype declaration
+        or None otherwise."""
+        # TODO: determine if this and get_typevar_declaration should be refactored
+        if len(s.lvalues) != 1 or not isinstance(s.lvalues[0], NameExpr):
+            return None
+        if not isinstance(s.rvalue, CallExpr):
+            return None
+        call = s.rvalue
+        if not isinstance(call.callee, RefExpr):
+            return None
+        callee = call.callee
+        if callee.fullname != 'typing.NewType':
+            return None
+        return call
+
+    def get_newtype_name(self, s: AssignmentStmt) -> Optional[str]:
+        lvalue = cast(NameExpr, s.lvalues[0])
```

Comment:
This cast scares me, because the check that ensure it's valid is in get_newtype_declaration(). I wonder if that function should just return a tuple with both the name and the call? (Then probably (None, None) on errors.)